### PR TITLE
Replaces old date picker for experience booking with a new calendar d…

### PR DESCRIPTION
**Pull request by benRmc**
…ate picker.

**Description**

**Issues that should be CLOSED by merge of this PR:**
* Fixes #

**Related issues that should be LINKED to this PR:**
* Connects #

### CODE REVIEW

File: traveler-ios-ui/TravelerKitUI/Storyboards/Wishlist.storyboard [hunk 1]
```diff
@@ -22,7 +22,7 @@
                     <simulatedNavigationBarMetrics key="simulatedTopBarMetrics" prompted="NO"/>
                     <connections>
                         <segue destination="Imj-5k-MIp" kind="custom" identifier="errorSegue" customClass="ContainerEmbedSegue" customModule="TravelerKitUI" id="4Rh-Tf-qWK"/>
-                        <segue destination="7yh-Lv-OwI" kind="custom" identifier="emptySegue" customClass="ContainerEmbedSegue" customModule="TravelerKitUI" id="Fu9-04-fSq"/>
+                        <segue destination="7yh-Lv-OwI" kind="custom" identifier="noResultSegue" customClass="ContainerEmbedSegue" customModule="TravelerKitUI" id="Fu9-04-fSq"/>
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Storyboards/Wishlist.storyboard]: Revert this.  Fixed in https://github.com/Guestlogix/traveler-ios/pull/326

File: traveler-ios-ui/TravelerKitUI/Storyboards/BookingItem.storyboard [hunk 2]
```diff
@@ -906,8 +772,348 @@
...
         </scene>
+        <!--Calendar Loading-->
+        <scene sceneID="BTF-L2-Eu6">
+            <objects>
+                <viewController title="Calendar Loading" id="htf-p3-PyO" sceneMemberID="viewController">
+                    <view key="view" contentMode="scaleToFill" id="sBA-I5-fCz">
+                        <rect key="frame" x="0.0" y="0.0" width="375" height="812"/>
+                        <autoresizingMask key="autoresizingMask" widthSizable="YES" heightSizable="YES"/>
+                        <subviews>
+                            <activityIndicatorView opaque="NO" contentMode="scaleToFill" horizontalHuggingPriority="750" verticalHuggingPriority="750" style="medium" translatesAutoresizingMaskIntoConstraints="NO" id="aaH-gI-m5g">
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/Storyboards/BookingItem.storyboard]: Note: UIActivityIndicatorView.Style.medium is iOS 13.0+

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 3]
```diff
@@ -14,14 +14,17 @@ public protocol AvailabilityViewControllerDelegate: class {
...
+    @IBOutlet weak var calendarContainer: UIView!
     @IBOutlet weak var priceLabel: UILabel!
     @IBOutlet weak var nextButton: UIButton!
-    @IBOutlet weak var tableView: UITableView!
 
     weak var delegate: AvailabilityViewControllerDelegate?
 
     var product: BookingItem?
-    var selectedAvailability: Availability?
+    private var selectedAvailability: Availability?
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: move down to rest of the private vars for organization

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 4]
```diff
@@ -14,14 +14,17 @@ public protocol AvailabilityViewControllerDelegate: class {
 }
 
 open class AvailabilityViewController: UIViewController {
+    @IBOutlet weak var previousMonthButton: UIButton!
+    @IBOutlet weak var monthYearLable: UILabel!
```
Comment (rhoque-glx) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: `monthYearLabel`

File: traveler-ios-ui/TravelerKitUI/Storyboards/BookingItem.storyboard (hunk 2 above)
Comment (rhoque-glx) [reply traveler-ios-ui/TravelerKitUI/Storyboards/BookingItem.storyboard]: Not sure of implications here, just a note.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 5]
```diff
@@ -88,112 +145,84 @@ open class AvailabilityViewController: UIViewController {
...
+        pageController?.delegate = self
+
+        addChild(pageController!)
+        pageController?.view.autoresizingMask = [.flexibleWidth,.flexibleHeight]
+        pageController?.view.frame = calendarContainer.bounds
+        calendarContainer.addSubview(pageController!.view)
+        pageController?.didMove(toParent: self)
+
+        let storyboard = UIStoryboard(name: "BookingItem", bundle: Bundle(for: AvailabilityViewController.self))
+        let initialVC = storyboard.instantiateViewController(withIdentifier: "calendarVC") as! AvailabilityCalendarViewController
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: The point of relying heavily in storyboards is to show how the flow behaves visually. Doing this defeats that purpose. If you're going to put a container view inside a view controller, then do all your setup and building  in the storyboard not here. Examples of this: Add flights, booking item details, etc. It is of paramount importance that you follow convention. The goal is for the codebase to look like it was written by a single person.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 6]
```diff
@@ -65,17 +85,54 @@ open class AvailabilityViewController: UIViewController {
         }
     }
 
+    @IBAction func didSelectPreviousMonth(_ sender: Any) {
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: This logic does not belong here. Currently this view controller is doing more than what it should. Availability view controller should only be in charge of fetching availabilities. Other view controllers should be in charge of dealing with calendar dates and selection.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 7]
```diff
@@ -31,22 +34,39 @@ open class AvailabilityViewController: UIViewController {
     }
     var selectedOption: BookingOption?
     var optionsViewController: BookingOptionsViewController?
+    var numberOfMonths = 12
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: This does not belong here! 1) Its bad practice to hard code values that are potentially configurable (like the number of months you want to display in a calendar) 2) This is calendar logic and it belongs in a different view controller.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 8]
```diff
@@ -31,22 +34,39 @@ open class AvailabilityViewController: UIViewController {
...
-
-    private var datePickerCellVisible = false
-    private var dateCellIndexPath: IndexPath {
-        return IndexPath(row: 0, section: 0)
-    }
-
-    private var datePickerCellIndexPath: IndexPath {
-        return IndexPath(row: 1, section: 0)
+    private var pageController: UIPageViewController?
+    private var calendarViewControllers: [Date: AvailabilityCalendarViewController] = [:]
```
Comment (OmarPadierna) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: Poor naming convention! it doesnt make sense to have a dictionary called "calendarViewControllers" where its items have "availabilityCalendarViewController"  Availability != (just) calendar. it should be "calendarViewControllers": [Data: CalendarViewController]. When your naming convention stops making sense, and your forced to mix stuff like this, it is normally an indication of two things 1) You're not naming things correctly 2) You've implemented a bad architecture and your controllers are doing too much. In this case it is the second.

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 9]
```diff
@@ -88,112 +145,84 @@ open class AvailabilityViewController: UIViewController {
...
+        guard let volatileDate = Calendar.current.date(byAdding: .month, value: 1, to: selectedDate) else {
+            Log("Unknown error getting the date of next month.", data: nil, level: .error)
+            return nil
+        }
 
-        present(alert, animated: true, completion: nil)
+        if let calendarVC = calendarViewControllers[volatileDate] {
+            return calendarVC
+        } else {
+            let storyboard = UIStoryboard(name: "BookingItem", bundle: Bundle(for: AvailabilityViewController.self))
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: You should never reference storyboards like this and instantiate VCs

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 10]
```diff
@@ -14,14 +14,17 @@ public protocol AvailabilityViewControllerDelegate: class {
 }
 
 open class AvailabilityViewController: UIViewController {
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: This VC is too bloated. It is not the VCs responsibility to handle the logic of a calendar display.

Make a new UIView subclass that encapsulates all calendar stuff and use it inside this VC..Keep VCs thin

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarViewController.swift [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+import UIKit
+import TravelerKit
+
+protocol AvailabilityCalendarViewControllerDelegate: class {
+    func availabilityCalendarViewController(_ controller: AvailabilityCalendarViewController, didSelect availability: Availability)
+}
+
+class AvailabilityCalendarViewController: UIViewController {
+    var representingDate: Date?
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarViewController.swift]: What is `representingDate`?

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarViewController.swift [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    var product: Product?
+    var availabilities: [Availability]?
+    var delegate: AvailabilityCalendarViewControllerDelegate?
+
+    override open func viewDidLoad() {
+        super.viewDidLoad()
+
+        if let _ = availabilities {
+            performSegue(withIdentifier: "resultSegue", sender: nil)
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarViewController.swift]: I dont understand this logic.

Either the VC is given the availabilites, or it fetches them..not both

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarResultViewController.swift [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+private let edgeInset = 10
+private let cellSpace = 5
+let calendarDateCellIdentifier = "calendarDateCellIdentifier"
+
+protocol AvailabilityCalendarResultViewControllerDelegate: class {
+    func availabilityCalendarResultViewController(_ controller: AvailabilityCalendarResultViewController, didSelect availability: Availability)
+}
+
+class AvailabilityCalendarResultViewController: UICollectionViewController {
+    var representingDate: Date?
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarResultViewController.swift]: ??

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarResultViewController.swift [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    var availabilities: [Availability]?
+    var delegate: AvailabilityCalendarResultViewControllerDelegate?
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarResultViewController.swift]: delegates MUST all be WEAK...this creates a memory leak.

Please read up on why this is the case

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarResultViewController.swift [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+//  TravelerKitUI
+//
+//  Created by Ben Ruan on 2019-10-28.
+//  Copyright © 2019 GuestLogix. All rights reserved.
+//
+
+import UIKit
+import TravelerKit
+
+private let edgeInset = 10
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityCalendarResultViewController.swift]: This is not an inset. Inset is its own type

Also these values dont belong here,...they need to be encapsulated inside a custom UIView and be configurable

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 16]
```diff
@@ -65,17 +85,54 @@ open class AvailabilityViewController: UIViewController {
...
+            Log("Unknown error getting the date of next month.", data: nil, level: .error)
+            return
+        }
+
+        selectedDate = volatileDate
+        if let calendarVC = calendarViewControllers[selectedDate] {
+            pageController?.setViewControllers([calendarVC], direction: .forward, animated: true, completion: nil)
+        } else {
+            let storyboard = UIStoryboard(name: "BookingItem", bundle: Bundle(for: AvailabilityViewController.self))
+            let calendarVC = storyboard.instantiateViewController(withIdentifier: "calendarVC") as! AvailabilityCalendarViewController
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: Do NOT use a PagingViewContorller

That VC is meant for finite set of VCs, not for piggybacking off of its swipe gesture

File: traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift [hunk 17]
```diff
@@ -88,112 +145,84 @@ open class AvailabilityViewController: UIViewController {
...
-
-            if selectedAvailability == nil {
-                updateSelectedDate(Date())
-
-                tableView.reloadRows(at: [indexPath], with: .none)
-            }
-        default:
-            break
+extension AvailabilityViewController: UIPageViewControllerDelegate {
+    public func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/ViewControllers/AvailabilityViewController.swift]: This is a misuse of paging view controller

File: traveler-ios-ui/TravelerKitUI/Views/CalendarDateCell.swift [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+//  Copyright © 2019 GuestLogix. All rights reserved.
+//
+
+import UIKit
+import TravelerKit
+
+class CalendarDateCell: UICollectionViewCell {
+    @IBOutlet weak var dateLabel: UILabel!
+
+    var date: Date? {
```
Comment (ata-n) [inline traveler-ios-ui/TravelerKitUI/Views/CalendarDateCell.swift]: This cell should be way dumber than knowing about dates and stuff.

It should ONLY have a label and selected and deselected states

File: traveler-swift-core/TravelerKit/Models/ProductItemCategory.swift [hunk 19]
```diff
@@ -17,4 +17,6 @@ public enum ProductItemCategory: String, Decodable {
     case show = "Show"
     case event = "Event"
     case nightlife = "Nightlife"
+    case NA = "NA"
```
Comment (ata-n) [inline traveler-swift-core/TravelerKit/Models/ProductItemCategory.swift]: NA is not acceptable..Maybe speak to backend guys to sort this out

**Pull request closed without merging**